src/Telemetry/Telemetry.cs

```diff
@@ -0,0 +1,157 @@
+            // Store the session ID in a static field so that it can be reused
+            CurrentSessionId = Guid.NewGuid().ToString();
+
+            string productVersion = typeof(Telemetry).Assembly.GetName().Version.ToString();
+            //initialize in task to offload to parallel thread
+            this._trackEventTask = Task.Factory.StartNew(() => this.InitializeTelemetry(productVersion));
+            this._initialized = true;
+        }
+
+        public bool Enabled { get; private set; }
+
+        public void TrackEvent(string eventName, IDictionary<string, string> properties,
+            IDictionary<string, double> measurements)
+        {
+            if (!this._initialized || !this.Enabled)
+            {
+                return;
+            }
+            this._logger.LogInformation($"Sending event {eventName}");
+
+            //continue task in existing parallel thread
+            this._trackEventTask = this._trackEventTask.ContinueWith(
+                x => this.TrackEventTask(eventName, properties, measurements)
+            );
+        }
+
+        private void InitializeTelemetry(string productVersion)
+        {
+            try
+            {
+                var config = new TelemetryConfiguration(InstrumentationKey);
+                this._client = new TelemetryClient(config);
+                this._client.Context.Session.Id = CurrentSessionId;
+                this._client.Context.Device.OperatingSystem = RuntimeInformation.OSDescription;
+
+                this._commonProperties = new TelemetryCommonProperties(productVersion).GetTelemetryCommonProperties();
+                this._commonMeasurements = new Dictionary<string, double>();
+            }
+            catch (Exception e)
+            {
+                this._client = null;
+                // we don't want to fail the tool if telemetry fails.
+                Debug.Fail(e.ToString());
+            }
+        }
+
+        private void TrackEventTask(
+            string eventName,
+            IDictionary<string, string> properties,
+            IDictionary<string, double> measurements)
+        {
+            if (this._client is null)
+            {
+                return;
+            }
+
+            try
+            {
+                Dictionary<string, string> eventProperties = this.GetEventProperties(properties);
+                Dictionary<string, double> eventMeasurements = this.GetEventMeasures(measurements);
+
+                this._client.TrackEvent($"{EventsNamespace}/{eventName}", eventProperties, eventMeasurements);
+                this._client.Flush();
+            }
+            catch (Exception e)
+            {
+                Debug.Fail(e.ToString());
+            }
+        }
+
+        private Dictionary<string, double> GetEventMeasures(IDictionary<string, double> measurements)
+        {
+            var eventMeasurements = new Dictionary<string, double>(this._commonMeasurements);
```

Comment:
Can we do any of these other solutions instead (to clean up this code)? https://stackoverflow.com/questions/6422091/convert-idictionary-to-dictionary

Reply:
What are you suggesting? Because there's two things happening here : 

1. We create a clone of the commonMeasurements dictionary (we don't want to modify the original)
2. We then add all the passed in measurements to that dictionary, overwriting any properties that might exist

We want to create a clone of the dictionary regardless (using the dictionary passed in could be problematic, we can't trust the caller to be doing the right thing there). And as far as I know there isn't a way to create a "combined" dictionary from two separate sources. 

Unless you're just saying that the logic is fine you just think we could compress it - in which case maybe? I'd have to play around with it, but that doesn't really seem worth the effort here given how short they are already.

Reply:
Oh, I missed line 128 where we're creating a clone of commonMeasurements. I thought this was just cloning in quite a roundabout fashion.

And yeah was mainly around compressing this into fewer lines. It's fine.